kernel/src/engine/arrow_expression.rs

```diff
@@ -135,6 +135,82 @@ fn column_as_struct<'a>(
         .ok_or(ArrowError::SchemaError(format!("{} is not a struct", name)))
 }
 
+fn make_arrow_error(s: String) -> Error {
+    Error::Arrow(arrow_schema::ArrowError::InvalidArgumentError(s))
+}
+
+/// Ensure a kernel data type matches an arrow data type. This only ensures that the actually "type"
+/// is the same, but does so recursively into structs, and ensures lists and maps have the correct
+/// associated types as well. This returns an `Ok` of the arrow type that can be used if the types
+/// match, or an error if the types do not match.
+fn ensure_data_types(
```

Comment:
This code seems to be stuck between two approaches:
* Doc comment and some code structure suggest it's massaging the type and returning the result
* All match arms actually just return `arrow_type.clone()`, and all call sites seem to just throw away non-error results as a way to validate recursively?

If we can get away with it, the validator approach seems cleaner?

Reply:
You're right. It's because I started it one way and realized it worked better as more of a validator. I think it probably makes sense to just have it return a `Result<()>` and use it purely as a validator. will update to that